adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,162 @@
+= Extension Structure Guidelines
+
+* Status: Proposed
+* Date: 2025-06-30
+* Authors: @cescoffier, @gsmet
+
+== Context
+
+The Quarkus ecosystem includes more than 700 extensions, many of which were developed without a consistent structure. The Quarkus core repository alone contains over 150 extensions.
+
+Most extensions follow the basic runtime/deployment split, but the boundaries between API and internal logic have often blurred. Additionally, without clear package structure rules, many split packages have emerged. This situation introduces several issues:
+
+* It is unclear which classes are intended as public APIs versus internal implementations.
+* There is no clear boundary between what an extension consumer can depend on and what they should not use.
+* Some extensions mix SPI and implementation code, reducing modularity and increasing the chance of unwanted coupling.
+* The need for build item re-use or coordination between extensions leads to ad-hoc patterns.
+
+More recently, we introduced dev-only dependencies, further increasing complexity.
+
+To address these issues and prepare for long-term maintainability (and possibly Java Platform Module System adoption), we propose a more structured layout for extension modules.
+
+== Decision
+
+Extensions may use a standardized structure composed of up to five optional modules, depending on their needs:
+
+* `runtime`: Contains the runtime implementation. It may expose public APIs if needed but should prefer delegation to the spi module.
+* `spi`: Contains public APIs intended to be consumed by other extensions or application code. This module should have minimal dependencies. Depending on a spi module does not transitively include the full extension (i.e., it avoids pulling in runtime or deployment logic).
+* `deployment`: Contains build-time logic, including BuildSteps and recorder logic. This module may define internal build items.
+* `deployment-spi`: Contains build-time APIs intended to be reused across multiple extensions. Extensions contributing to or using build items from other extensions should depend on deployment-spi modules.
+* `runtime-dev`: Contains runtime classes used exclusively in dev mode (e.g., for the Dev UI). This avoids shipping dev-only classes into production artifacts.
+
+All these modules are optional.
+
+=== Additional Rules
+
+	* Module dependencies must follow strict rules:
+	* `deployment` depends on `runtime`, `runtime-dev` (if defined) and `deployment-spi` (if defined).
```

Comment:
`deployment` does not need to depend on `runtime-dev`. It may be necessary but there is no general reason for that.